contracts/FlowContractAudits.cdc

```diff
@@ -0,0 +1,215 @@
+        }
+
+        pub fun deleteVoucher(key: String) {
+            self.auditorCapability!.borrow()!.deleteVoucher(key: key)
+        }
+
+        init() {
+            self.auditorCapability = nil
+        }
+
+    }
+
+    // Can be called by anyone but needs a capability to function
+    pub fun createAuditorProxy(): @AuditorProxy {
+        return <- create AuditorProxy()
+    }
+
+    pub resource Administrator {
+        
+        // Creates new Auditor
+        pub fun createNewAuditor(): @Auditor {
+            emit AuditorCreated()
+            return <-create Auditor()
+        }
+
+        // Checks all vouchers and removes expired ones
+        pub fun cleanupExpiredVouchers() {
+            for key in FlowContractAudits.vouchers.keys {                
+                let v = FlowContractAudits.vouchers[key]!
+                if v.expiryBlockHeight != nil {
+                    if getCurrentBlock().height > v.expiryBlockHeight! {
+                        FlowContractAudits.vouchers.remove(key: key)
+                        emit VoucherRemoved(key: key, recurrent: v.recurrent, expiryBlockHeight: v.expiryBlockHeight)                
+                    }
+                }
+            }
+        }
+
+        // For testing
+        pub fun useVoucherForDeploy(address: Address, code: String): Bool {
+            return FlowContractAudits.useVoucherForDeploy(address: address, code: code)
+        }
+    }
+
+    // This function will be called by the FVM on contract deploy/update
+    access(contract) fun useVoucherForDeploy(address: Address, code: String): Bool {
+        let codeHash = FlowContractAudits.hashContractCode(code)
+        var key = FlowContractAudits.generateVoucherKey(address: address, codeHash: codeHash)
+
+        // first check for voucher based on target account
+        // if not found check for any account
+        if !FlowContractAudits.vouchers.containsKey(key) {
+            key = FlowContractAudits.generateVoucherKey(address: nil, codeHash: codeHash)
+            if !FlowContractAudits.vouchers.containsKey(key) {
+                return false
+            }
+        }
+
+        let v = FlowContractAudits.vouchers[key]!
+
+        // ensure contract code matches the voucher
+        if v.codeHash == codeHash  {
+
+            // if expiryBlockHeight is set, check the current block height
+            // and remove/expire the voucher if not within the acceptable range
+            if v.expiryBlockHeight != nil {
+                if getCurrentBlock().height > v.expiryBlockHeight! {
+                    FlowContractAudits.vouchers.remove(key: key)
+                    emit VoucherRemoved(key: key, recurrent: v.recurrent, expiryBlockHeight: v.expiryBlockHeight)
+                    return false
+                }
+            }
+
+            // remove the voucher if not recurrent
+            if !v.recurrent {
+                FlowContractAudits.vouchers.remove(key: key)
```

Comment:
It might be better to define a `removeVoucher` function that removes the voucher from the dictionary and emits the event. This way it might be harder to forget to emit the event if you are removing a voucher.

This is just a suggestion though. It is also fine as it is.

Reply:
I agree, this code piece has been repeated multiple times. Here's my take at refactoring: https://github.com/onflow/flow-core-contracts/pull/271/commits/8d7a1e5bda44a3372b4fb2ae1542b2e6a99961ed